translator/trace/protospan_to_jaegerthrift_test.go

```diff
@@ -0,0 +1,245 @@
+	tracepb "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
+	"github.com/census-instrumentation/opencensus-service/internal/testutils"
+	"github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/jaegertracing/jaeger/thrift-gen/jaeger"
+)
+
+func TestTraceIDRoundTripJaegerOCProtoJaeger(t *testing.T) {
+	wl := int64(0x0001020304050607)
+	wh := int64(0x70605040302010FF)
+	gl, gh := traceIDBytesToLowAndHigh(jTraceIDToOCProtoTraceID(wh, wl))
+	if gl != wl || gh != wh {
+		t.Errorf("Round trip of trace Id failed want: (0x%0x, 0x%0x) got: (0x%0x, 0x%0x)", wl, wh, gl, gh)
+	}
+}
+
+func TestSpanIDRoundTripJaegerOCProtoJaeger(t *testing.T) {
+	w := int64(0x0001020304050607)
+	g := ocIDBytesToJaegerID(jSpanIDToOCProtoSpanID(w))
+	if g != w {
+		t.Errorf("Round trip of span Id failed want: 0x%0x got: 0x%0x", w, g)
+	}
+}
+
+func TestOCProtoToJaegerThrift(t *testing.T) {
+	const numOfFiles = 2
+	for i := 0; i < numOfFiles; i++ {
+		ocBatch := ocBatches[i]
+
+		gotJBatch, err := OCProtoToJaegerThrift(ocBatch)
+		if err != nil {
+			t.Errorf("Failed to translate OC batch to Jaeger Thrift: %v", err)
+			continue
+		}
+
+		wantSpanCount, gotSpanCount := len(ocBatch.Spans), len(gotJBatch.Spans)
+		if wantSpanCount != gotSpanCount {
+			t.Errorf("Different number of spans in the batches on pass #%d (want %d, got %d)", i, wantSpanCount, gotSpanCount)
+			continue
+		}
+
+		// Jaeger binary tags do not round trip from Jaeger -> OCProto -> Jaeger.
+		// For tests use data without binary tags.
+		thriftFile := fmt.Sprintf("./testdata/thrift_batch_no_binary_tags_%02d.json", i+1)
+		wantJBatch := &jaeger.Batch{}
+		if err := loadFromJSON(thriftFile, wantJBatch); err != nil {
+			t.Errorf("Failed load Jaeger Thrift from %q: %v", thriftFile, err)
+			continue
+		}
+
+		// Sort tags to help with comparison, not only for jaeger.Process but also
+		// on each span.
+		sort.Slice(gotJBatch.Process.Tags, func(i, j int) bool {
+			return gotJBatch.Process.Tags[i].Key < gotJBatch.Process.Tags[j].Key
+		})
+		sort.Slice(wantJBatch.Process.Tags, func(i, j int) bool {
+			return wantJBatch.Process.Tags[i].Key < wantJBatch.Process.Tags[j].Key
+		})
+		var jSpans []*jaeger.Span
+		jSpans = append(jSpans, gotJBatch.Spans...)
+		jSpans = append(jSpans, wantJBatch.Spans...)
+		for _, jSpan := range jSpans {
+			sort.Slice(jSpan.Tags, func(i, j int) bool {
+				return jSpan.Tags[i].Key < jSpan.Tags[j].Key
+			})
+		}
+
+		gjson, _ := json.Marshal(gotJBatch)
+		wjson, _ := json.Marshal(wantJBatch)
+		gjsonStr := testutils.GenerateNormalizedJSON(string(gjson))
+		wjsonStr := testutils.GenerateNormalizedJSON(string(wjson))
+		if gjsonStr != wjsonStr {
+			t.Errorf("OC Proto to Jaeger Thrift failed.\nGot:\n%s\nWant:\n%s\n", gjsonStr, wjsonStr)
+			t.Errorf("Saving formated jsons at ./testdata/ to help with further investigation")
+			testutils.SaveAsFormattedJSON(fmt.Sprintf("./testdata/failure_oc_to_jaeger_got_%02d.json", i), gotJBatch)
```

Comment:
I don't think we want tests writing arbitrary JSON to file systems. Please remove this unless we clear these files on exit but this might not be pleasant to users who clone our project and get a bug.

Reply:
If a change breaks the test this should speedup the fixing it. It will be visible as untracked file together with the other test files.

Reply:
I am uncomfortable with writing arbitrary files and leaving them on user's disks. In many projects I've encountered this as a discomfort and actually last week, I helped clean up such in the Go project from a stray write https://github.com/golang/go/issues/28781. We can print them to stdout.

Reply:
Think of these files as penalty for someone breaking the tests 😃 , but that is fine, I can quickly add this code if I want to compare the files. I will remove it and the related function.

Reply:
LOLOL